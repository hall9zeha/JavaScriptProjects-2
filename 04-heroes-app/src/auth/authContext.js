import { createContext } from "react";

//Como la devolución de createContext() es un componente, el nombre de la constante 
//tendrá la sintaxis de un Componente(con mayúsculas por convención), esto es opcional claro
export const AuthContext = createContext()