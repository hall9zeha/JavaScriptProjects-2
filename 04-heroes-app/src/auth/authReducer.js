import { Types } from "../types/Types";

export const authReducer = (state ={}, action)=>{

    switch (action.type) {
        case Types.login:
                return {
                    ...action.payload,
                    logged:true
                }
        case Types.logout:
            return {
                logged:false
            }
               
        default:
            return state;
    }
}