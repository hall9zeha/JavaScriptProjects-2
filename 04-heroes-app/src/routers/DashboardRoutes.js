import { Route, Routes } from "react-router-dom"
import { DCScreen } from "../components/dc/DCScreen"
import { MarvelScreen } from "../components/marvel/MarvelScreen"
import { SearchScreen } from "../components/search/SearchScreen"
import { Navbar } from "../components/ui/NavBar"
import { HeroScreen } from "../components/hero/HeroScreen"

export const DashboardRoutes = () => {
  return (
    <>
        {/* Para mostrar solo el navbar  cuando se navegue a los siguiente componentes */}
        <Navbar/>
        <div className="container">
        <Routes>
            <Route path='/' element={<MarvelScreen/>}/>
            <Route path='marvel' element={<MarvelScreen/>}/>
            <Route path='dc' element={<DCScreen/>}/>
            <Route path='search' element={<SearchScreen/>}/>
            <Route path='hero/:heroId' element={<HeroScreen/>}/>
            
        </Routes>
        </div>
    </>
  )
}
