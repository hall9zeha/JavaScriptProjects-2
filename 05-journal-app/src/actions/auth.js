// Las acciones no son más que simples funciones que representan nuestros casos de uso

import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/Types";
import { uiFinishLoading, uiStartLoading } from "./ui";



export const startLoginWithEmailPassword = (email, password) =>{
    return (dispatch)=>{
        dispatch(uiStartLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName));
                dispatch(uiFinishLoading())
            })
            .catch(e=>{
                dispatch(uiFinishLoading())
                console.log(e)
            })
        
    }
}

export const startLoginWithGoogle = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName))
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async({user}) =>{
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid,user.displayName));
            })
            .catch(e=>{
                console.log(e)
            })
    }
}


// La siguiente función retornará un objeto
export const login = (uid, displayName)=>({
    type: types.login,
    payload:{
        uid,
        displayName
    }
})

// Logout

export const startLogout = ()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout=()=>({
    type:types.logout
})