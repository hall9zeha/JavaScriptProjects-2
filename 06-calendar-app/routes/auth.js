/*
User routes /auth
host + /api/auth

*/

const {Router} = require('express');
const router = Router();
const {check} = require('express-validator')

const {createUser, login,renewToken} = require('../controllers/auth');
const { fieldValidate } = require('../middlewares/field-validators');
const { validateJWT } = require('../middlewares/jws-validator');

router.post('/new',
    //Midlewares para validar los campos necesarios, comenzamos a hacerlos aqu√≠ y los recibimos en nuestros controllers

    [
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'The password must have 6 or more characters').isLength({min:6}),
        fieldValidate
    ],    
    createUser)
router.post('/',
    [
        check('email', 'Email is required').isEmail(),
        check('password', 'The password must have 6 or more characters').isLength({min:6}),
        fieldValidate
    ],
    login)
    // Validamos el json web token con vlidateJWT 
router.get('/renew',validateJWT, renewToken)

module.exports = router;