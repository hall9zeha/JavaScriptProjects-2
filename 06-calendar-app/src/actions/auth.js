import Swal from "sweetalert2";
import { fetchWithToken, fetchWithoutToken } from "../helpers/fetch"
import { types } from "../types/types";

export const startLogin = (email, password) =>{
    //Gracias a la librería thunk podemos disponer de un dispatch
    return async(dispatch) =>{
        const resp = await fetchWithoutToken('auth',{email,password},'POST');
        const body = await resp.json();
        
        //Si recibimos un arespuesta correcta de nuestra api 
        /*
            {
                ok:true
                ...
            }
        guardamos el token y la fecha de creación en local storage
        */
       
        if(body.ok){
            localStorage.setItem('token',body.token);
            localStorage.setItem('token-init-date',new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))
        }else{
            Swal.fire('Error',body.msg, 'error');
        }
    }
}

export const startRegister = (name,email,password) =>{
    return async(dispatch)=>{
        const resp = await fetchWithoutToken('auth/new',{name,email,password},'POST');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
        }

    }
}
export const startChecking =()=> {
    return async(dispatch)=>{

        //No requiere ningún argumento porque por defecto será una petición GET
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
            dispatch(checkingFinish());
        }
    }
}

const checkingFinish = () =>({type:types.authCheckingFinish});
//Ya que usaremos esta acción solo en este lugar no la exportaremos
const login = (user) =>({
    type:types.authLogin,
    payload:user
})

