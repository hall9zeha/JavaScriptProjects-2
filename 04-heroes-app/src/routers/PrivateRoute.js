import React, {useContext } from "react"
import { AuthContext } from "../auth/authContext"
import { Navigate, useLocation } from "react-router-dom";

//Muy importante el nombre del objeto a desescructurar es 'children' 
export const PrivateRoute = ({children}) => {
    const location = useLocation();
    const {user} = useContext(AuthContext);
    const {pathname, search} = location;

    //Guardamos la última dirección visitada para retornar a ella cuando se haga login
    localStorage.setItem('lastPath', pathname + search);

  return user.logged ? children : <Navigate to="/login"/>
}
