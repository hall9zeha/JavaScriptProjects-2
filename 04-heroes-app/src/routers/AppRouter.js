
import React from 'react'
import { BrowserRouter, Route, Routes } from 'react-router-dom'

import { LoginScreen } from '../components/login/LoginScreen'

import { DashboardRoutes } from './DashboardRoutes'

export const AppRouter = () => {
  return (
    <BrowserRouter>
        {/* Mostraremos la pantalla de login pero sin el navbar */}
        <Routes>
            <Route path='/login' element={<LoginScreen/>}/>
            <Route path='/*' element={<DashboardRoutes/>}/>
        </Routes>
    </BrowserRouter>
  )
}
