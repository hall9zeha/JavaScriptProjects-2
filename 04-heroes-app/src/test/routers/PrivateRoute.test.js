import { MemoryRouter, BrowserRouter as Router,Routes,Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { AuthContext } from '../../auth/authContext';
import { PrivateRoute } from '../../routers/PrivateRoute';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    Navigate: () => <span>Redireccionando</span>
  }));

describe('Test in <PrivateRoute/> component', () => { 

    Storage.prototype.setItem = jest.fn();

    test('should show the component if is authenticated and save in localStorage',()=>{
        const contextValue = {
            user:{
                logged:true,
                name:'Barry'
            }
        }

        const wrapper = renderer.create(
            <AuthContext.Provider value = {contextValue}>
                <MemoryRouter initialEntries={['/']}>
                    <PrivateRoute>
                        <h1>Private component</h1>
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.root.findByType('h1').children[0].trim()).toBe('Private component')
        expect(localStorage.setItem).toHaveBeenCalledWith("lastPath","/")

    })
    test('should lock the component if not auth', () => { 
        const contextValue = {
            user:{
                logged:false
            }
        }

        const wrapper = renderer.create(
            <AuthContext.Provider value = {contextValue}>
                <MemoryRouter initialEntries={['/']}>
                    <PrivateRoute>
                        <h1>Private component</h1>
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.root.findByType('span').children[0].trim()).toBe('Redireccionando')
     })
 })