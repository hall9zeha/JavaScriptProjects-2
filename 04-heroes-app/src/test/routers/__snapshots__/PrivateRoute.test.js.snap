// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <PrivateRoute/> component should lock the component if not auth 1`] = `
<span>
  Redireccionando
</span>
`;

exports[`Test in <PrivateRoute/> component should show the component if is authenticated and save in localStorage 1`] = `
<h1>
  Private component
</h1>
`;
